plonk/src/bencher.rs

```diff
@@ -0,0 +1,184 @@
+//! Helper functions for micro-benchmarks
+
+use ark_std::{thread_local, time::Instant};
+use core::{cell::RefCell, time::Duration};
+
+thread_local!(static FFT_START_TIME: RefCell<Instant> = RefCell::new(Instant::now()));
+thread_local!(static FFT_TIMER_LOCK: RefCell<bool> = RefCell::new(false));
+thread_local!(static FFT_TOTAL_TIME: RefCell<Duration> = RefCell::new(Duration::ZERO));
+
+thread_local!(static MSM_START_TIME: RefCell<Instant> = RefCell::new(Instant::now()));
+thread_local!(static MSM_TIMER_LOCK: RefCell<bool> = RefCell::new(false));
+thread_local!(static MSM_TOTAL_TIME: RefCell<Duration> = RefCell::new(Duration::ZERO));
+
+thread_local!(static POLY_EVAL_START_TIME: RefCell<Instant> = RefCell::new(Instant::now()));
+thread_local!(static POLY_EVAL_TIMER_LOCK: RefCell<bool> = RefCell::new(false));
+thread_local!(static POLY_EVAL_TOTAL_TIME: RefCell<Duration> = RefCell::new(Duration::ZERO));
+
+/// Initializing the timers
+#[inline]
+pub fn init_timers() {
+    #[cfg(feature = "bench")]
+    {
+        FFT_TOTAL_TIME.with(|timer| {
+            *timer.borrow_mut() = Duration::ZERO;
+        });
+        FFT_TIMER_LOCK.with(|lock| {
+            *lock.borrow_mut() = false;
+        });
+        MSM_TOTAL_TIME.with(|timer| {
+            *timer.borrow_mut() = Duration::ZERO;
+        });
+        MSM_TIMER_LOCK.with(|lock| {
+            *lock.borrow_mut() = false;
+        });
+        POLY_EVAL_TOTAL_TIME.with(|timer| {
+            *timer.borrow_mut() = Duration::ZERO;
+        });
+        POLY_EVAL_TIMER_LOCK.with(|lock| {
+            *lock.borrow_mut() = false;
+        });
+    }
+}
+
+/// Get the total time that we have spend on FFT related computations
+#[inline]
+pub fn total_fft_time() -> Duration {
+    #[cfg(feature = "bench")]
+    {
+        FFT_TOTAL_TIME.with(|duration| *duration.borrow())
+    }
+    #[cfg(not(feature = "bench"))]
+    Duration::ZERO
+}
+
+/// Get the total time that we have spend on MSM related computations
+#[inline]
+pub fn total_msm_time() -> Duration {
+    #[cfg(feature = "bench")]
+    {
+        MSM_TOTAL_TIME.with(|duration| *duration.borrow())
+    }
+    #[cfg(not(feature = "bench"))]
+    Duration::ZERO
+}
+
+/// Get the total time that we have spend on polynomial evaluations
+#[inline]
+pub fn total_poly_eval_time() -> Duration {
+    #[cfg(feature = "bench")]
+    {
+        POLY_EVAL_TOTAL_TIME.with(|duration| *duration.borrow())
+    }
+    #[cfg(not(feature = "bench"))]
+    Duration::ZERO
+}
+
+#[inline]
+pub(crate) fn fft_start() {
+    #[cfg(feature = "bench")]
+    {
+        if FFT_TIMER_LOCK.with(|lock| *lock.borrow()) {
+            panic!("another FFT timer has already started somewhere else");
+        }
+
+        FFT_START_TIME.with(|timer| {
+            *timer.borrow_mut() = Instant::now();
+        });
+
+        FFT_TIMER_LOCK.with(|lock| {
+            *lock.borrow_mut() = true;
+        })
```

Comment:
this logic is slightly baffling to me, so if FFT is locked, then we panic?
1. it seem to me that our timer will never be used by 2 threads at the same time? because the multi-threaded code is in-between the `fft_start()` and `fft_end()` (e.g. inside `fn compute_selector_polynomials`), so maybe we don't need this locking to begin with?
2. If we have proper concurrency control, then FFT timer should never panic, but rather wait on lock to be released instead?